### PR TITLE
gpu: Add CmdCopyBufferToImage

### PR DESCRIPTION
**Pull request by arno-lunarg**
Add 07931:
> If [VK_EXT_depth_range_unrestricted](https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/VK_EXT_depth_range_unrestricted.html) is not enabled, for each element of pRegions whose imageSubresource contains a depth aspect, the data in srcBuffer must be in the range [0,1]

Simplified a bit LogValidationMessage with the addition of
LogCustomValidationMessage.
While at it, removed a redundant gpu memory mapping.

Redo management of shared command validation resources:
Added a shared validation resources map to store them,
shared_validation_resources_map.
this map is cleared at PreCallRecordDestroyDevice time.
=> No need to connect new shared resources types by hand
to the Validation class.
Moved shared resources types inside the relevant classes.

### CODE REVIEW

File: layers/gpu_validation/gpu_subclasses.h [hunk 1]
```diff
@@ -207,7 +220,24 @@ class PreTraceRaysResources : public CommandResources {
 
     void Destroy(gpuav::Validator &validator) final;
     bool LogValidationMessage(gpuav::Validator &validator, VkQueue queue, VkCommandBuffer cmd_buffer, const uint32_t *debug_record,
-                              const uint32_t operation_index, const LogObjectList &objlist);
+                              const uint32_t operation_index, const LogObjectList &objlist) final;
+};
+
+class PreCopyBufferToImageResources : public CommandResources {
```
Comment (spencer-lunarg) [inline layers/gpu_validation/gpu_subclasses.h:226]: I am going to assume we will want a future `ImageToBuffer` and maybe `Image-to-Image` and `Buffer-to-buffer` and would probably like to group them together

so would suggest going towards a `Resource Copy` name scheme

Comment (arno-lunarg) [reply layers/gpu_validation/gpu_subclasses.h:213]: Resources needed in those other cases may be different, but if experience shows that they are the same I will rally them under the same banner

File: layers/gpu_validation/gpu_error_message.cpp [hunk 2]
```diff
@@ -527,49 +527,38 @@ bool gpuav::Validator::AnalyzeAndGenerateMessages(VkCommandBuffer cmd_buffer, Vk
...
     if (result == VK_SUCCESS) {
-        const uint32_t total_words = debug_output_buffer[spvtools::kDebugOutputSizeOffset];
+        const uint32_t total_words = output_buffer[spvtools::kDebugOutputSizeOffset];
         // A zero here means that the shader instrumentation didn't write anything.
         if (total_words != 0) {
-            uint32_t *debug_record = &debug_output_buffer[spvtools::kDebugOutputDataOffset];
+            uint32_t *debug_record = &output_buffer[spvtools::kDebugOutputDataOffset];
             const LogObjectList objlist(queue, cmd_buffer);
-            LogValidationMessage(validator, queue, cmd_buffer, debug_record, operation_index, objlist);
+            LogValidationMessage(validator, queue, cmd_buffer, output_buffer, debug_record, operation_index, objlist);
```
Comment (spencer-lunarg) [inline layers/gpu_validation/gpu_error_message.cpp:538]: so this is the only caller to `LogValidationMessage` now, so I would not pass in `debug_record` here, and call `uint32_t *debug_record = &output_buffer[spvtools::kDebugOutputDataOffset];` inside the function

File: layers/gpu_validation/gpu_image_layout.cpp [hunk 3]
```diff
@@ -345,46 +345,81 @@ void gpuav::Validator::PreCallRecordCmdCopyBufferToImage(VkCommandBuffer command
...
+
+    auto copy_buffer_to_image =
+        AllocatePreCopyBufferToImageValidationResources(record_obj.location.function, commandBuffer, &copy_buffer_to_image_info);
+    StoreCommandResources(commandBuffer, std::move(copy_buffer_to_image));
 }
 
 void gpuav::Validator::PreCallRecordCmdCopyBufferToImage2KHR(VkCommandBuffer commandBuffer,
                                                              const VkCopyBufferToImageInfo2KHR *pCopyBufferToImageInfo2KHR,
                                                              const RecordObject &record_obj) {
     BaseClass::PreCallRecordCmdCopyBufferToImage2KHR(commandBuffer, pCopyBufferToImageInfo2KHR, record_obj);
```
Comment (spencer-lunarg) [inline layers/gpu_validation/gpu_image_layout.cpp:384]: this should just call `PreCallRecordCmdCopyBufferToImage2` as we have `RecordObject` to know if it from the core or KHR version now

Comment (spencer-lunarg) [reply layers/gpu_validation/gpu_image_layout.cpp:384]:
```suggestion
    PreCallRecordCmdCopyBufferToImage(commandBuffer, pCopyBufferToImageInfo2KHR, record_obj);
```

File: layers/gpu_validation/gpu_validation.cpp [hunk 4]
```diff
@@ -975,160 +993,224 @@ std::unique_ptr<gpuav::CommandResources> gpuav::Validator::AllocatePreTraceRaysV
...
+    DispatchCmdDispatch(cmd_buffer, group_count_x, 1, 1);
+
+    // Restore the previous compute pipeline state.
+    restorable_state.Restore(cmd_buffer);
+
+    return copy_buffer_to_img_resources;
 }
 
-void gpuav::Validator::AllocateSharedDrawIndirectValidationResources(bool use_shader_objects) {
+gpuav::PreDrawResources::SharedResources *gpuav::Validator::GetSharedDrawIndirectValidationResources(bool use_shader_objects) {
```
Comment (spencer-lunarg) [inline layers/gpu_validation/gpu_validation.cpp:1204]: nit (honeslty happy to do in another PR too) but we should just use a `namespace` in this file as everything here is `gpuav`

File: layers/gpu_validation/gpu_validation.cpp [hunk 5]
```diff
@@ -975,160 +993,224 @@ std::unique_ptr<gpuav::CommandResources> gpuav::Validator::AllocatePreTraceRaysV
...
+        shared_resources != shared_validation_resources_map.end()) {
+        return reinterpret_cast<gpuav::PreDrawResources::SharedResources *>(shared_resources->second.get());
+    }
+
+    auto shared_resources = std::make_unique<gpuav::PreDrawResources::SharedResources>();
+
     VkResult result;
-    if (!common_draw_resources.initialized) {
+
         std::vector<VkDescriptorSetLayoutBinding> bindings = {
```
Comment (spencer-lunarg) [inline layers/gpu_validation/gpu_validation.cpp:1214]: this seems to still be indented, idk why clang-format didn't adjust it, but please un-indent this

File: layers/gpu_validation/gpu_validation.cpp [hunk 6]
```diff
@@ -1169,112 +1251,388 @@ void gpuav::Validator::AllocateSharedDrawIndirectValidationResources(bool use_sh
...
                 ReportSetupProblem(device, "Unable to create shader module. Aborting GPU-AV");
                 aborted = true;
-                return;
+                return nullptr;
             }
         }
 
-        common_draw_resources.initialized = true;
-    }
+        const auto elt =
```
Comment (spencer-lunarg) [inline layers/gpu_validation/gpu_validation.cpp:1275]: what does `elt` stand for?

File: layers/vulkan/generated/inst_functions_comp.h [hunk 7]
```diff
@@ -25,7 +25,7 @@
 ****************************************************************************/
 
 // To view SPIR-V, copy contents of array and paste in https://www.khronos.org/spir/visualizer/
-static const uint32_t inst_functions_comp[3042] = {
```
Comment (spencer-lunarg) [inline layers/vulkan/generated/inst_functions_comp.h:28]: why did this file change?

File: tests/unit/gpu_av.cpp [hunk 8]
```diff
@@ -904,3 +904,269 @@ TEST_F(NegativeGpuAV, DISABLED_YcbcrDrawFetchIndexed) {
     m_commandBuffer->EndRenderPass();
     m_commandBuffer->end();
 }
+
```
Comment (spencer-lunarg) [inline tests/unit/gpu_av.cpp:907]: I would like 1 positive test that does a `CmdCopyBufferToImage` of the depth buffer with valid input to make sure 100% we don't start spamming false positive by accident

File: layers/gpu_validation/gpu_subclasses.h (hunk 1 above)
Comment (spencer-lunarg) [reply layers/gpu_validation/gpu_subclasses.h:213]: sounds good, happy for now

File: layers/gpu_validation/gpu_validation.cpp [hunk 9]
```diff
@@ -975,160 +993,224 @@ std::unique_ptr<gpuav::CommandResources> gpuav::Validator::AllocatePreTraceRaysV
...
-void gpuav::Validator::AllocateSharedTraceRaysValidationResources() {
-    if (!common_trace_rays_resources.initialized) {
-        VkResult result = VK_SUCCESS;
-
-        std::vector<VkDescriptorSetLayoutBinding> bindings = {
-            {0, VK_DESCRIPTOR_TYPE_STORAGE_BUFFER, 1, VK_SHADER_STAGE_RAYGEN_BIT_KHR, nullptr}  // output buffer
-        };
+std::unique_ptr<gpuav::CommandResources> gpuav::Validator::AllocatePreCopyBufferToImageValidationResources(
+    vvl::Func cmd, VkCommandBuffer cmd_buffer, const VkCopyBufferToImageInfo2 *copy_buffer_to_img_info) {
+    // No need to perform validation if VK_EXT_depth_range_unrestricted is enabled
```
Comment (spencer-lunarg) [inline layers/gpu_validation/gpu_validation.cpp:998]: we have a `if (!gpuav_settings.validate_indirect_buffer) {` for the other calls

I would like to have a `validate_copies` in the `GpuAVSetting` to add this as a toggle-able item

File: layers/gpu_validation/gpu_image_layout.cpp (hunk 3 above)
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_image_layout.cpp:384]: ok yeah much nicer

File: layers/gpu_validation/gpu_validation.cpp (hunk 5 above)
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_validation.cpp:1214]: yeah wtf

File: layers/gpu_validation/gpu_validation.cpp (hunk 4 above)
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_validation.cpp:1204]: yesssss I should have done it ^^;

File: layers/gpu_validation/gpu_validation.cpp (hunk 6 above)
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_validation.cpp:1275]: element

File: layers/vulkan/generated/inst_functions_comp.h (hunk 7 above)
Comment (arno-lunarg) [reply layers/vulkan/generated/inst_functions_comp.h:28]: Because it `#include "gpu_shaders_constants.h"`

File: tests/unit/gpu_av.cpp (hunk 8 above)
Comment (arno-lunarg) [reply tests/unit/gpu_av.cpp:907]: Yes it is safer

File: layers/gpu_validation/gpu_validation.cpp (hunk 9 above)
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_validation.cpp:998]: ok, will dig out how to do this again (￣_,￣ )

**Review by spencer-lunarg — APPROVED** [review]
LGTM - will need a rebase now

**Review by jeremyg-lunarg — COMMENTED** [review]
code LGTM, just some comments about the comments

File: layers/gpu_validation/gpu_subclasses.h [hunk 10]
```diff
@@ -135,7 +92,7 @@ struct AccelerationStructureBuildValidationInfo {
 
 class Validator;
 
-// Every recorded command need the validation resources listed in this function
+// Every recorded action command needs the validation resources listed in this function
```
[the remark this answers is not in the document]
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_subclasses.h:95]: I first heard Spencer refer to the draw/dispatch/trace rays commands as "action commands". I do not know if there is an official term to designate those kinds of command that essetially "execute" a bound pipeline?

File: layers/gpu_validation/gpu_subclasses.h [hunk 11]
```diff
@@ -148,10 +105,13 @@ class CommandResources {
 
     void LogErrorIfAny(gpuav::Validator &validator, VkQueue queue, VkCommandBuffer cmd_buffer, const uint32_t operation_index);
     // Return true iff an error has been logged
```
[the remark this answers is not in the document]
Comment (arno-lunarg) [reply layers/gpu_validation/gpu_subclasses.h:107]: removed

File: tests/unit/gpu_av.cpp [hunk 12]
```diff
@@ -972,3 +972,269 @@ TEST_F(NegativeGpuAV, DISABLED_YcbcrDrawFetchIndexed) {
...
+    VkBufferImageCopy buffer_image_copy_2 = buffer_image_copy_1;
+    buffer_image_copy_2.imageOffset = {32, 32, 0};
+    buffer_image_copy_2.imageExtent = {32, 32, 1};
+
+    vk::CmdCopyBufferToImage(*m_commandBuffer, copy_src_buffer, copy_dst_image, VK_IMAGE_LAYOUT_TRANSFER_DST_OPTIMAL, 1,
+                             &buffer_image_copy_2);
+
+    m_commandBuffer->end();
+    m_errorMonitor->SetDesiredFailureMsg(kErrorBit, "has a float value at offset 16376 that is not in the range [0, 1]");
```
[the remark this answers is not in the document]
Comment (arno-lunarg) [reply tests/unit/gpu_av.cpp:1032]: is there a good way to do that? Since in the complete error message there the VUID and this excerpt are separated by printed values like the buffer handle that vary, I cannot just copy paste the entire thing

File: layers/gpu_validation/gpu_subclasses.h (hunk 10 above)
[the remark this answers is not in the document]
Comment (artem-lunarg) [reply layers/gpu_validation/gpu_subclasses.h:95]: That's the term used in the specification for commands that do something on the gpu

>  In general, action commands are those commands that alter framebuffer attachments, read/write buffer or image memory, or write to query pools.

[the remark this answers is not in the document]
Comment (spencer-lunarg) [reply layers/gpu_validation/gpu_subclasses.h:95]: https://github.com/KhronosGroup/Vulkan-Docs/issues/1908

I tried, it really will need to be a VVL defined name at this point
